### PR TITLE
cephadm: compile cephadm

### PR DESCRIPTION
**Pull request by p-se**
Compiling cephadm into an single binary offers the ability to split the code from cephadm up into several files/modules. A side effect of having a `cephadm.py` file (which, by renaming to cephadm.py has already become a module) is that it can properly be imported and static type checking inside `src/cephadm/tests/test_cephadm.py` is not an issue anymore.

Next steps would to split up cephadm into several modules and adapting the test_cephadm.py import mechanism, as well as fixing the typing issues, which are revealed by using a proper important statement.

**edit**

It even enabled `cephadm.py` (new name) to be checked by mypy. I didn't notice mypy didn't check the file correctly until it had that `.py` suffix. That's why I disabled that check for this PR, but follow-up PRs should handle resolve those issues.

Signed-off-by: Patrick Seidensal <pseidensal@suse.com>

# TODO

- [x] Use zipapp for file creation
- [x] Ensure cehpadm  and vstart work correctly
- [x] Fix/test packages (rpm/deb)
- [ ] Find/implement solution for more or less direct download of compile cephadm binary

See also:

* #32526

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/cephadm/cephadm.py
Comment (p-se) [inline src/cephadm/cephadm.py:2]: This ignores errors that have previously been ignored by mypy because `cephadm` (without the `.py` extension) wasn't a proper Python module.

Comment (mgfritch) [reply src/cephadm/cephadm.py:2]: this has bothered me too. when we re-enable this, it would also be nice to refactor the type checking from py2 -> py3 style syntax

**Comment by mgfritch** [comment]
> Next steps would to split up cephadm into several modules and adapting the test_cephadm.py import mechanism, as well as fixing the typing issues, which are revealed by using a proper important statement.

I'm in favor of idea(s) to allow us to better organize bin/cephadm!

**Review by mgfritch — COMMENTED** [review]
`make check` needs a rebase against  #41854 and  #41859?

Also, have we addressed https://github.com/ceph/ceph/pull/41614#discussion_r644663183? I think we will need a solution that allows for both a user to pull the complied binary as well as the QA suite to test against a binary built from a pending PR ...

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
I think the hardest part for this PR is finding a good replacement for 

* https://github.com/ceph/ceph/blob/462c1216535b9466571cf2cda8ea2adb4fc179df/doc/cephadm/install.rst#curl-based-installation
* https://github.com/ceph/ceph/blob/a67d1cf2a7a4031609a5d37baa01ffdfef80e993/doc/install/get-packages.rst#L30

while replacing this:

https://github.com/ceph/ceph/blob/9ab9cc26e200cdc3108525770353b91b3dd6c6d8/qa/tasks/cephadm.py#L126-L140

would be a lot easier: it's just code.

**Comment by p-se** [comment]
The way cephadm gets packed might change a bit, but the good news is that there's a nice way to pack external dependencies. They'll have to be defined in a requirements.txt file and can be downloaded automatically using pip, then everything is packed. I've created a demo in case you're curious.

https://github.com/p-se/python-zipapp

The demo bundles all dependencies required to run fastapi. One downside seems to be the file size. Even compressed it still makes it to 14M. cephadm will probably need to think twice about which dependencies are being added. Note that it seems not to be possible to bundle any dependencies based on compiled languages.

File: src/cephadm/CMakeLists.txt
Comment (tchaikov) [inline src/cephadm/CMakeLists.txt:7]: presumably, we should redistribute the "compiled" cephadm which is built from the dist tarball. so i think we should 

1. create the temporary files and `cephadm` in `${CMAKE_BINARY_DIR}` not `${CMAKE_SOURCE_DIR}`
2. install the built `cephadm` instead. i.e. update `debian/rules` and `ceph.spec.in` accordingly. currently, we just copy `src/cephadm/cephadm` to `sbin/cephadm`. after the change, we should install it using `install(PROGRAMS ...)` in cmake.

**Comment by tchaikov** [comment]
https://github.com/ceph/ceph/blob/a67d1cf2a7a4031609a5d37baa01ffdfef80e993/doc/install/get-packages.rst#L30

how about
```sh
curl <where-ceph-adm-is> | sh
```

just like https://rustup.rs .

**Comment by sebastian-philipp** [comment]
> how about
> 
> ```shell
> curl <where-cephadm-is> | sh
> ```
> 
> just like https://rustup.rs .

That wouldn't solve our problem as we need the binary to work with. Piping into sh won't really help us for the cephadm binary.

**Comment by tchaikov** [comment]
> That wouldn't solve our problem as we need the binary to work with. Piping into sh won't really help us for the cephadm binary.

hmm, i admit that i don't understand why it's difficult. as the shell script could be any shell script which bootstrap the necessary bits, which in turn downloads the python scripts or even use `pip` to install a python package from a github repo.

File: src/cephadm/CMakeLists.txt
Comment (p-se) [reply src/cephadm/CMakeLists.txt]: Thank you!

**Comment by sebastian-philipp** [comment]
```
[  3%] compiling /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/arm64/AVAILABLE_ARCH/arm64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-7506-g2f1144a7/rpm/el8/BUILD/ceph-17.0.0-7506-g2f1144a7/aarch64-redhat-linux-gnu/bin/cephadm
usage: zipapp.py [-h] [--output OUTPUT] [--python PYTHON] [--main MAIN]
                 [--info]
                 source
zipapp.py: error: unrecognized arguments: --compress
make[2]: *** [src/cephadm/CMakeFiles/cephadm.dir/build.make:82: bin/cephadm] Error 2
make[1]: *** [CMakeFiles/Makefile2:7421: src/cephadm/CMakeFiles/cephadm.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=arm64,AVAILABLE_ARCH=arm64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/57288/consoleFull#-85362248744e9240e-b50a-4693-bac0-8a991bac86ac

https://shaman.ceph.com/builds/ceph/wip-swagner3-testing-2021-09-07-1225/

**Comment by p-se** [comment]
> ```
> [  3%] compiling /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/arm64/AVAILABLE_ARCH/arm64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-7506-g2f1144a7/rpm/el8/BUILD/ceph-17.0.0-7506-g2f1144a7/aarch64-redhat-linux-gnu/bin/cephadm
> usage: zipapp.py [-h] [--output OUTPUT] [--python PYTHON] [--main MAIN]
>                  [--info]
>                  source
> zipapp.py: error: unrecognized arguments: --compress
> make[2]: *** [src/cephadm/CMakeFiles/cephadm.dir/build.make:82: bin/cephadm] Error 2
> make[1]: *** [CMakeFiles/Makefile2:7421: src/cephadm/CMakeFiles/cephadm.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ```
> 
> https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=arm64,AVAILABLE_ARCH=arm64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/57288/consoleFull#-85362248744e9240e-b50a-4693-bac0-8a991bac86ac
> 
> https://shaman.ceph.com/builds/ceph/wip-swagner3-testing-2021-09-07-1225/

I see, `--compress` was introduced in 3.7. That's unfortunate, as compression does a good job in reducing the file size. If that becomes a problem, we might want to fall back to compression of SSH or so.

**Comment by p-se** [comment]
@tchaikov I needed to check which Python version is used for zipping cephadm, so that I didn't have to remove the `--compress` argument completely. Unfortunately, I'm not at all confident to do that with CMakeLists.txt, so I switched back to the `build.sh` script and adapted it.

The dependencies you've mentioned earlier and which are installed by `src/cephadm/build.sh` if the `requirements.txt` file is present is not required in this pull request, that is true. However, a requirement for this implementation is that it will be possible to add external dependencies to be packaged into the `cephadm` file. This is why the `build.sh` script already supports that.

**Comment by tchaikov** [comment]
@p-se sorry for the latency, will reply you late today.

File: src/pybind/mgr/cephadm/ssh.py
Comment (tchaikov) [inline src/pybind/mgr/cephadm/ssh.py:144]: per the type annotation, `stdin` can only be `bytes` or `None`. if that's the case, i'd prefer putting something like
```py
if stdin is None:
  encoding = `utf-8`
else:
  encoding = None
r = await conn.run(cmd, input=stdin, encoding=encoding)
```

for two reasons:

1. to be more explicit
2. `stdin` is None or a `bytes` object, so there is no need to put `stdin if stdin else None`.

File: src/pybind/mgr/cephadm/ssh.py
Comment (tchaikov) [inline src/pybind/mgr/cephadm/ssh.py:156]: nit, i'd prefer check `encoding` instead of dispatching by the type of an object.

**Comment by tchaikov** [comment]
> @tchaikov I needed to check which Python version is used for zipping cephadm, so that I didn't have to remove the `--compress` argument completely. Unfortunately, I'm not at all confident to do that with CMakeLists.txt, so I switched back to the `build.sh` script and adapted it.

sure. i will help on this part once your change is approved and merged.

> 
> The dependencies you've mentioned earlier and which are installed by `src/cephadm/build.sh` if the `requirements.txt` file is present is not required in this pull request, that is true. However, a requirement for this implementation is that it will be possible to add external dependencies to be packaged into the `cephadm` file. This is why the `build.sh` script already supports that.

ack.

**Comment by sebastian-philipp** [comment]
https://shaman.ceph.com/builds/ceph/wip-swagner3-testing-2021-09-13-1702/

File: qa/tasks/cephadm.py
Comment (sebastian-philipp) [inline qa/tasks/cephadm.py:129]: flake8: `./tasks/cephadm.py:10:1: F401 're' imported but unused`

**Comment by sebastian-philipp** [comment]
```
+ install -m 0755 src/cephadm/cephadm /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-7645-g4f8f0957/rpm/el8/BUILDROOT/ceph-17.0.0-7645.g4f8f0957.el8.x86_64/usr/sbin/cephadm
install: cannot stat 'src/cephadm/cephadm': No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.YQt85I (%install)
```

File: qa/tasks/cephadm.py
Comment (mgfritch) [reply qa/tasks/cephadm.py:129]: fixed during the rebase

File: src/pybind/mgr/cephadm/ssh.py
Comment (mgfritch) [reply src/pybind/mgr/cephadm/ssh.py]: maybe it would be simpler to interact with asyncssh using [raw bytes](https://asyncssh.readthedocs.io/en/latest/api.html#asyncssh.SSHClientConnection.create_session)?

```
... `None` can be passed in if the application wishes to send and receive raw bytes
```

File: src/pybind/mgr/cephadm/ssh.py
Comment (mgfritch) [reply src/pybind/mgr/cephadm/ssh.py]: .. like above simply assume raw bytes?

**Comment by mgfritch** [comment]
jenkins test make check

File: src/pybind/mgr/cephadm/ssh.py
Comment (tchaikov) [reply src/pybind/mgr/cephadm/ssh.py]: after a second thought. yeah, that'd be a lot simpler!

**Comment by tchaikov** [comment]
jenkins test signed

File: src/pybind/mgr/cephadm/ssh.py
Comment (mgfritch) [reply src/pybind/mgr/cephadm/ssh.py]: errr actually I don't think this will work, because some callers expect to pass input/output as str data (that will be utf-8 encode/decode) and yet others expect raw binary bytes data (see `_write_remote_file`).

transferring remote files (cephadm zip binary) via`asyncssh.scp` might make this implementation a bit cleaner: 
https://github.com/ceph/ceph/pull/43458

**Comment by mgfritch** [comment]
Jenkins retest this please

File: qa/tasks/cephadm.py
Comment (kshtsk) [inline qa/tasks/cephadm.py:134]: Try to get the target arch from the context and any remote node where you are going to run the image.

Comment (kshtsk) [reply qa/tasks/cephadm.py:134]: Also, keep in mind that `arch` record can be provided within the config and if it is ignore, someone who gonna refactor this code in future may want to know why.

**Comment by phlogistonjohn** [comment]
FYI - @mgfritch is going to take some time off and in the meantime I plan on helping with this effort. :wave: 

As discussed earlier today the plan is first to make the current build.sh more pythonic and work on any test failures.

**Comment by mgfritch** [comment]
a few failures to track down from the latest test run:

https://pulpito.ceph.com/mgfritch-2022-06-09_20:03:05-orch:cephadm-wip-mgfritch-testing-2022-06-09-1029-distro-default-smithi/

specifically the upgrade test failures:
```
 2022-06-10T01:44:56.244 INFO:teuthology.orchestra.run.smithi153.stderr:__main__.Error: Container release octopus != cephadm release quincy; please use matching version of cephadm (pass --allow-mismatched-release to continue anyway)
```

**Comment by phlogistonjohn** [comment]
jenkins test dashboard
jenkins test api

**Comment by phlogistonjohn** [comment]
jenkins test dashboard cephadm

**Comment by phlogistonjohn** [comment]
jenkins test make check

**Comment by phlogistonjohn** [comment]
jenkins test make check

File: qa/tasks/cephadm.py
Comment (anthonyeleven) [inline qa/tasks/cephadm.py:146]: Suggest "Install cephadm from an RPM package"

Very minor.

**Review by anthonyeleven — CHANGES_REQUESTED** [review]

File: qa/tasks/cephadm.py
Comment (phlogistonjohn) [reply qa/tasks/cephadm.py]: Sure. I tweaked it to be more accurate in the process. Thanks!

**Review by anthonyeleven — APPROVED** [review]

**Comment by anthonyeleven** [comment]
Note that my approval is only for docs, not for code

File: qa/tasks/cephadm.py
Comment (dmick) [inline qa/tasks/cephadm.py:7]: "chacra" is misspelled in the commit message; should fix before merge

**Review by dmick — CHANGES_REQUESTED** [review]

**Comment by phlogistonjohn** [comment]
https://pulpito.ceph.com/phlogistonjohn-2022-07-14_19:47:09-orch:cephadm-wip-phlogistonjohn-testing-2022-07-12-1005-distro-default-smithi/

All tests for orch/cephadm have passed in this run. The plan is now to also test some of the other suites that make use of cephadm.
CC: @mgfritch @adk3798

**Comment by phlogistonjohn** [comment]
https://pulpito.ceph.com/phlogistonjohn-2022-07-15_14:35:27-fs:upgrade-wip-phlogistonjohn-testing-2022-07-12-1005-distro-default-smithi/
fs/upgrade - All passed.

**Comment by phlogistonjohn** [comment]
https://pulpito.ceph.com/phlogistonjohn-2022-07-18_17:10:24-rados:cephadm-wip-phlogistonjohn-testing-2022-07-12-1005-distro-default-smithi/
1 dead, the remainder passed  - assuming dead job is a fluke and moving on

**Comment by phlogistonjohn** [comment]
https://pulpito.ceph.com/phlogistonjohn-2022-07-18_18:30:41-upgrade-wip-phlogistonjohn-testing-2022-07-12-1005-distro-default-smithi/
A higher proportion of failures here but nothing seems directly connected to cephadm. Possibly related to rgw's  https://tracker.ceph.com/issues/55853 and jumping from octopus->main in some of the upgrade tests.

**Comment by phlogistonjohn** [comment]
I'm pretty much happy with the approach to running teuthology tests with complied cephadm OR an older script version depending on the YAML content. This should work for now but we will probably want to revisit it once we need upgrade tests *from* a version that complies cephadm.

As a  next step I want to rebase again and also clean up the commit history. I personally think that some of the patches would benefit from being squashed. In this case that would mean squashing patches of different authors so, to be polite, I'd like to ask if that's OK.  How about that @mgfritch ?  I'll add co-author tags to the squashed commits as needed.

File: qa/tasks/cephadm.py
Comment (phlogistonjohn) [reply qa/tasks/cephadm.py]: Fixed, thanks.

**Comment by mgfritch** [comment]
> As a next step I want to rebase again and also clean up the commit history. I personally think that some of the patches would benefit from being squashed. In this case that would mean squashing patches of different authors so, to be polite, I'd like to ask if that's OK. How about that @mgfritch ? I'll add co-author tags to the squashed commits as needed.

@phlogistonjohn  seems reasonable to me!

**Comment by phlogistonjohn** [comment]
I've rebased and squashed some of the patches. I'm generally happy with the state of the patches. I think it's time for the next round of feedback and then final testing with intent to merge!

File: qa/tasks/util/chacra.py
Comment (SvenMW) [inline qa/tasks/util/chacra.py:176]: you may want to set correct permissions when writing a binary file.

File: qa/tasks/cephadm.py
Comment (rkachach) [inline qa/tasks/cephadm.py:163]: you may consider using f-strings instead of old formatting

File: qa/tasks/cephadm.py
Comment (rkachach) [inline qa/tasks/cephadm.py:253]: this seems repetitive code, maybe we can have it a function?

File: qa/tasks/util/chacra.py
Comment (rkachach) [inline qa/tasks/util/chacra.py:182]: We are always returning with 0 regardless of any error that could happen while downloading or writing the file!

File: qa/tasks/cephadm.py
Comment (phlogistonjohn) [reply qa/tasks/cephadm.py:163]: Thanks, but this is not net-new code it's just moving what was already in the file. I'd rather save changes like this for future PRs.

File: qa/tasks/util/chacra.py
Comment (phlogistonjohn) [reply qa/tasks/util/chacra.py:182]: Because this is a `main()` function the expectation is that if this line in reached then no exceptions were raised and the return code of zero is passed to sys.exit. While I don't think it's strictly necessary, I don't think it's wrong either.

File: qa/tasks/cephadm.py
Comment (phlogistonjohn) [reply qa/tasks/cephadm.py:253]: That would probably be good for a follow up pr.

File: qa/tasks/util/chacra.py
Comment (phlogistonjohn) [reply qa/tasks/util/chacra.py:176]: Thanks for pointing that out, however I think the permissions are being adjusted later in the code. It could make a good future enhancement. However, since this PR is long and complex I don't want to make that change now since the tests *are* passing.

**Review by adk3798 — APPROVED** [review]

**Comment by adk3798** [comment]
Additional runs of rados/cephadm and some fs suite tests that use cephadm
https://pulpito.ceph.com/adking-2022-09-14_12:51:53-rados:cephadm-wip-adk-testing-2022-09-13-1304-distro-default-smithi/
https://pulpito.ceph.com/adking-2022-09-14_12:49:36-fs-wip-adk-testing-2022-09-13-1304-distro-default-smithi/

**Comment by sebastian-philipp** [comment]
Congrats @adk3798 @p-se @mgfritch @rkachach @phlogistonjohn   for getting this merged. :tada:

File: qa/tasks/cephadm.py
Comment (rzarzynski) [inline qa/tasks/cephadm.py:171]: Does this code really work? For `main` and `reef` I'm getting:

```
$ curl https://raw.githubusercontent.com/ceph/ceph/main/src/cephadm/cephadm
404: Not Found
```

Comment (phlogistonjohn) [reply qa/tasks/cephadm.py:171]: The `download_cephadm` function calls either _fetch_cephadm_from_github (what you highlighted) or _fetch_cephadm_from_chachra*. The former should generally be called when teuthololgy needs to check out an old version of cephadm. The latter should be called when a newer cephadm is needed.

There's a comment in `download_cephadm` that says `TODO: come up with a sensible way to detect if we need an "old, uncompiled" cephadm`. It is possible you are hitting a case where the detection is not right.  Maybe file a tracker issue and CC me (and Adam King) and I'll try to help some more when I return to work next week.



*I just noticed this is also a typo.

Comment (rzarzynski) [reply qa/tasks/cephadm.py:171]: Hi @phlogistonjohn!

The problem has popped up during the reef2squid upgrade testing: https://pulpito.ceph.com/yuriw-2023-11-23_15:43:47-upgrade:reef-x-wip-yuri10-testing-2023-11-22-1112-distro-default-smithi/.

All the jobs in the `upgrade:reef-x` suite failed due to:

```
023-11-23T16:06:23.673 DEBUG:teuthology.orchestra.run.smithi119:> sudo hostname $(hostname -s)
2023-11-23T16:06:23.700 INFO:tasks.cephadm:Downloading cephadm (repo https://github.com/ceph/ceph ref reef)...
2023-11-23T16:06:23.701 DEBUG:teuthology.orchestra.run.smithi114:> curl --silent https://raw.githubusercontent.com/ceph/ceph/reef/src/cephadm/cephadm > /home/ubuntu/cephtest/cephadm && ls -l /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:23.923 INFO:teuthology.orchestra.run.smithi114.stdout:-rw-rw-r--. 1 ubuntu ubuntu 14 Nov 23 16:06 /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:23.924 DEBUG:teuthology.orchestra.run.smithi119:> curl --silent https://raw.githubusercontent.com/ceph/ceph/reef/src/cephadm/cephadm > /home/ubuntu/cephtest/cephadm && ls -l /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:24.018 INFO:teuthology.orchestra.run.smithi119.stdout:-rw-rw-r--. 1 ubuntu ubuntu 14 Nov 23 16:06 /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:24.018 DEBUG:teuthology.orchestra.run.smithi114:> test -s /home/ubuntu/cephtest/cephadm && test $(stat -c%s /home/ubuntu/cephtest/cephadm) -gt 1000 && chmod +x /home/ubuntu/cephtest/cephadm
2023-11-23T16:06:24.036 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

Replicated locally:

```
$ curl --silent https://raw.githubusercontent.com/ceph/ceph/reef/src/cephadm/cephadm
404: Not Found
```

For quincy it works.

I tracked the problem to 18fcfb022d19af1e69addf1729d4593272e4ac20 which got merged as a part of this PR.

There was a follow-up but the fix is imperfect: https://github.com/ceph/ceph/pull/48232.

Comment (adk3798) [reply qa/tasks/cephadm.py:171]: @rzarzynski the first commit from https://github.com/ceph/ceph/pull/53105 was intended to address pulling the cephadm binary from stable branches where it is compiled (currently only reef). I had the PR on hold becuase reef -> main upgrades currently don't work anyway because main is version 18.0.0 and reef is 18.2.0 so cephadm considers it a downgrade and blocks the upgrade. I know you were working on the v19 kickoff stuff. What I can probably do is separate out the first commit of that PR and add some additional work to have the reef-x upgrade tests make use of it. Then we could test that with the v19 kickoff work when it's ready.

**Comment by dmick** [comment]
I'm not certain it's relevant to this PR, but I wanted to make sure y'all were aware of this code in ceph-build:

https://github.com/ceph/ceph-build/blob/main/scripts/build_utils.sh#L989-L993
https://github.com/ceph/ceph-build/blob/main/ceph-dev-new-build/build/build_rpm#L59-L63

It's asymmetric because build_deb has been pulled into generic code while build_rpm is still in each job's build scripts, but the idea is the same in both cases.  The advantage here is that the build does whatever it does to compile and button up cephadm, and this just consumes the result.  (it also places it in a findable location in the chacra repo, not as a package, but just an ancillary file, for copying directly from there when testing).

I did that code; I don't know if any of the testing code actually consumes cephadm from that location.

File: qa/tasks/cephadm.py
Comment (adk3798) [reply qa/tasks/cephadm.py:171]: PR for what I mentioned in previous comment https://github.com/ceph/ceph/pull/54721

**Pull request merged**
